src/include/brain/kd_tree.h

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// kd_tree.h
+//
+// Identification: src/include/brain/kd_tree.h
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "common/macros.h"
+#include "annoy/annoylib.h"
+#include "annoy/kissrandom.h"
+#include "brain/cluster.h"
+
+namespace peloton {
+namespace brain {
+
+//===--------------------------------------------------------------------===//
+// KDTree
+//===--------------------------------------------------------------------===//
+
+class KDTree {
+ public:
+  KDTree(int num_features)
+      : num_features_(num_features), index_(num_features) {}
+
+  void Insert(Cluster *cluster) {
+    index_.unbuild();
+    index_.add_item(size_, cluster->GetCentroid().data());
+    index_.build(2 * num_features_);
```

Comment:
So even with insert we still need to completely rebuild the kd-tree? I thought it's only that we cannot update or delete... This still may not be the bottleneck, but I just want to confirm.